src/libmongoc/src/mongoc/mongoc-cmd.c

```diff
@@ -474,6 +474,13 @@ _mongoc_cmd_parts_assemble_mongos (mongoc_cmd_parts_t *parts,
       hedge = mongoc_read_prefs_get_hedge (parts->read_prefs);
    }
 
+   if (server_stream->effective_read_mode != (mongoc_read_mode_t) 0) {
+      /* Server selection may have overriden the read mode used to generate this
+       * server stream. This has effects on the body of the message that we send
+       * to the server */
+      mode = server_stream->effective_read_mode;
```

Comment:
`mongoc_read_mode_t` is not a complete read preference on its own. Isn't this problematic if the "effective read preference" is more complex (e.g. includes tags) and we need to send _that_ in the outgoing command?

Reply:
The read mode is only tracked here so that we can later decide whether to send the read preference at all. The command builder will have access to the full read preference data, but it will only need to know whether server selection ignored the read mode or not.

Reply:
Ah, thanks for clearing that up. I'd be in favor of changing this to a boolean in the interest of clarity/maintainability if that's not too much trouble.

Reply:
> if and only if the `effective_read_mode` is primary or 0 [...] only need to know whether server selection ignored the read mode or not

I am also in favor of the boolean.

Suggestion: I recommend `must_use_primary` as the name per wording in the [Server Selection Spec](https://github.com/mongodb/specifications/blob/3a4a6e0f318545c9f050a0000fbd53fc4d9a2790/source/server-selection/server-selection.rst#use-of-read-preferences-with-commands):
> For pre-5.0 servers, an aggregate command is "must-use-primary" if its pipeline contains a write stage (e.g. `$out`, `$merge`) ...

Reply:
There is more than just a true/false state on how server selection behaved. I don't know how to narrow this down to a single bit of information and keep the same behavior. It may be possible, but I'm not sure what it would look like.

Reply:
> There is more than just a true/false state on how server selection behaved. I don't know how to narrow this down to a single bit of information and keep the same behavior. It may be possible, but I'm not sure what it would look like.

IIUC the information needed in `mongoc_cmd_parts_assemble` is "Can the user's read preference be applied, or should it be overridden with a primary read preference".

This is what I have in mind for the logic below:

```
   mode = mongoc_read_prefs_get_mode (prefs_ptr);
   if (server_stream->must_use_primary) {
      /* Server selection determined the user's read preference must be overridden with primary */
      mode = MONGOC_READ_PRIMARY;
   }
```